File: solutions/226. Invert Binary Tree/226.js
// Method 1: Recursive, DFS
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var invertTree = function(root) {
    if (root == null) return null;

    let left = root.left;
    let right = root.right;

    root.left = invertTree(right);
    root.right = invertTree(left);

    return root;
};

// Method 2: Iterative using Queue, BFS
/*
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
 /*
var invertTree = function(root) {
    let queue = [root];

    while (queue[0]) {
        let node = queue.shift();
        let left = node.left;
        node.left = node.right;
        node.right = left;

        if (node.left) {
            queue.push(node.left);
        }
        if (node.right) {
            queue.push(node.right);
        }
    }

    return root;
};
*/

File: neetcode/226. Invert Binary Tree/226.kt
/*
Method 1.0:
Intuition: Attach the inverted subtrees to opposite sides of root node
TC: O(N), N is the number of nodes in the tree
SC: O(1)
 */
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun invertTree(root: TreeNode?): TreeNode? {
        if (root == null) {
            return null
        }

        val left = root.left
        val right = root.right

        root.left = invertTree(right)
        root.right = invertTree(left)

        return root
    }
}
